### PR TITLE
LocalChannel write when peer closed leak

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
If LocalChannel doWrite executes while the peer's state changes from CONNECTED to CLOSED it is possible that some promise's won't be completed and buffers will be leaked.

Modifications:
- Check the peer's state in doWrite to avoid a race condition

Result:
All write operations should release, and the associated promise should be completed.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@normanmaurer @nmittler @ejona86 - FYI

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java [hunk 1]
```diff
@@ -306,7 +322,7 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
         case BOUND:
             throw new NotYetConnectedException();
         case CLOSED:
-            throw new ClosedChannelException();
+            throw CLOSED_CHANNEL_EXCEPTION;
```
Comment (nmittler) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: Why this change? Is the stacktrace not useful?

**Comment by nmittler** [comment]
@Scottmitch LGTM

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java (hunk 1 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: Not that it isn't useful, just more consistent with other channels (see [AbstractChannel](https://github.com/netty/netty/blob/master/transport/src/main/java/io/netty/channel/AbstractChannel.java#L793)).

Comment (nmittler) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: Yup ... it's fine :)

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java [hunk 2]
```diff
@@ -241,12 +255,17 @@ protected void doClose() throws Exception {
...
+                try {
+                    peer.eventLoop().execute(new OneTimeTask() {
+                        @Override
+                        public void run() {
+                            doPeerClose(peer, peerWriteInProgress);
+                        }
+                    });
+                } catch (RuntimeException e) {
+                    // The peer close may attempt to drain this.inboundBuffers. If that fails make sure it is drained.
+                    releaseInboundBuffers();
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: @Scottmitch should we re-throw at least or log ?

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java [hunk 3]
```diff
@@ -295,7 +314,11 @@ protected void doBeginRead() throws Exception {
                 threadLocals.setLocalChannelReaderStackDepth(stackDepth);
             }
         } else {
-            eventLoop().execute(readTask);
+            try {
+                eventLoop().execute(readTask);
+            } catch (RuntimeException e) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: same as above

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java [hunk 4]
```diff
@@ -357,10 +386,24 @@ public void run() {
...
-        } else {
-            peer.eventLoop().execute(finishPeerReadTask);
+        try {
+            if (peer.writeInProgress) {
+                peer.finishReadFuture = peer.eventLoop().submit(finishPeerReadTask);
+            } else {
+                peer.eventLoop().execute(finishPeerReadTask);
+            }
+        } catch (RuntimeException cause) {
+            peer.releaseInboundBuffers();
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: same as above

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java (hunk 2 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: Good catch. The intention was to re-throw.

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java (hunk 3 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: rethrow

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java (hunk 4 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: rethrow

**Comment by Scottmitch** [comment]
Cherry-picked 4.0 (7130837) 4.1 (1e763b6) master (1031b78)

**Comment by Scottmitch** [comment]
Also master (3084482) (woops)
